File: src/com/leozhi/topic88/Solution03.java
package com.leozhi.topic88;

/**
 * @author leozhi
 * 双指针
 * 通过
 * 0ms
 */
public class Solution03 {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        int i = m - 1, j = n - 1, index = nums1.length - 1;
        while (i >= 0 || j >= 0) {
            if (i < 0) {
                while (j >= 0) {
                    nums1[index] = nums2[j];
                    index--;
                    j--;
                }
                break;
            }
            if (j < 0) {
                break;
            }
            if (nums1[i] > nums2[j]) {
                nums1[index] = nums1[i];
                index--;
                i--;
            } else {
                nums1[index] = nums2[j];
                index--;
                j--;
            }
        }
    }
}

File: src/com/leozhi/topic496/Solution03.kt
package com.leozhi.topic496

import java.util.*
import kotlin.collections.HashMap

/**
 * @author leozhi
 * 单调栈
 * 通过
 * 216ms
 */
class Solution03 {
    fun nextGreaterElement(nums1: IntArray, nums2: IntArray): IntArray {
        val res = IntArray(nums1.size)
        val stack = Stack<Int>()
        // key 为 nums2 中的元素，value 为各元素下一个更大的数
        val map = HashMap<Int, Int>()
        for (num in nums2) {
            while (!stack.empty() && num > stack.peek()) {
                map[stack.pop()] = num
            }
            stack.push(num)
        }
        // 此时栈中剩余的元素其后没有更大的数
        while (!stack.empty()) {
            map[stack.pop()] = -1
        }
        for (i in nums1.indices) {
            res[i] = map[nums1[i]]!!
        }
        return res
    }
}